scipy/special/_convex_analysis.pxd

```diff
@@ -20,11 +21,26 @@ cdef inline double kl_div(double x, double y) noexcept nogil:
     else:
         return INFINITY
 
+@cython.cdivision(True)
 cdef inline double rel_entr(double x, double y) noexcept nogil:
+    cdef double ratio
+    # Smallest non-subnormal number, aka np.finfo(np.float64).tiny
+    cdef double tiny = 2.2250738585072014e-308
     if isnan(x) or isnan(y):
         return NAN
     elif x > 0 and y > 0:
-        return x * log(x / y)
+        ratio = x / y
+        if 0.5 < ratio < 2:
+            # When x and y are close, this is more accurate
+            return x * log1p((x - y) / y)
+        if tiny < ratio < INFINITY:
+            # There are no underflow/overflow issues
+            return x * log(ratio)
+        else:
+            # x and y are so far apart that taking x / y
+            # results in either an underflow, overflow,
+            # or subnormal number. Do the logarithm first
+            return x * (log(x) - log(y))
     elif x == 0 and y >= 0:
         return 0
     else:
```

Comment:
GitHub isn't letting me turn this into a suggestion, but I think it would be good to avoid most of the nested indentation. You could structure this as

```cython
    if isnan(x) or isnan(y):
        return NAN
    if x <= 0 or y <= 0:
        if x == 0 and y >= 0:
            return 0
        return INFINITY
    ratio = x / y
    if 0.5 < ratio < 2:
        # When x and y are close, this is more accurate
        return x * log1p((x - y) / y)
    if DBL_MIN < ratio < INFINITY:
        # There are no underflow/overflow issues
        return x * log(ratio)
    # x and y are so far apart that taking x / y
    # results in either an underflow, overflow,
    # or subnormal number. Do the logarithm first
    return x * (log(x) - log(y))
```

I replaced `tiny` with `DBL_MIN` following my other suggestion.

Reply:
Fixed in 3727337.